xtext.cpp

```diff
@@ -0,0 +1,2157 @@
+#define MAX_FONT_NUM 16
+
+#define NOT_MEASURED INT_MIN
+#define REPLACEMENT_CHAR 0xFFFD
+#define ELLIPSIS_CHAR 0x2026
+
+// Helpers with font metrics (units are 1/64 px)
+#define FONT_METRIC_FLOOR(x)    ((x) & -64)
+#define FONT_METRIC_CEIL(x)     (((x)+63) & -64)
+#define FONT_METRIC_ROUND(x)    (((x)+32) & -64)
+#define FONT_METRIC_TRUNC(x)    ((x) >> 6)
+
+// Uncomment for debugging text measurement and line shaping:
+// #define DEBUG_MEASURE_TEXT
+// #define DEBUG_SHAPE_LINE
+
+// ==============================================
+// Utility functions
+
+inline bool is_unicodepoint_rtl(uint32_t c) {
+    // Try to detect if this unicode codepoint is a RTL char
+    // Looking at fribidi/lib/bidi-type.tab.i and its rules for tagging
+    // a char as RTL, only the following ranges will trigger it:
+    //   0590>08FF      Hebrew, Arabic, Syriac, Thaana, Nko, Samaritan...
+    //   200F 202B      Right-To-Left mark/embedding control chars
+    //   202E 2067      Right-To-Left override/isolate control chars
+    //   FB1D>FDFF      Hebrew and Arabic presentation forms
+    //   FE70>FEFF      Arabic presentation forms
+    //   10800>10FFF    Other rare scripts possibly RTL
+    //   1E800>1EEBB    Other rare scripts possibly RTL
+    // (There may be LTR chars in these ranges, but we're ok with false
+    // positives: we'll invoke fribidi, which will say there's no bidi.)
+    // Try to balance the searches
+    bool is_rtl = false;
+    if ( c >= 0x0590 ) {
+        if ( c <= 0x2067 ) {
+            if ( c <= 0x08FF ) is_rtl = true;
+            else if ( c >= 0x200F ) {
+                if ( c == 0x200F || c == 0x202B || c == 0x202E || c == 0x2067 ) is_rtl = true;
+            }
+        }
+        else if ( c >= 0xFB1D ) {
+            if ( c <= 0xFDFF ) is_rtl = true;
+            else if ( c <= 0xFEFF ) {
+                if ( c >= 0xFE70) is_rtl = true;
+            }
+            else if ( c <= 0x1EEBB ) {
+                if (c >= 0x1E800) is_rtl = true;
+                else if ( c <= 0x10FFF && c >= 0x10800 ) is_rtl = true;
+            }
+        }
+    }
+    return is_rtl;
+}
+
+// Fribidi provides fribidi_charset_to_unicode(FRIBIDI_CHAR_SET_UTF8,...)
+// but it expects valid utf8, and we want to support broken UTF-8 and WTF-8.
+// So we implement Utf8ToUnicode(), which can be called twice:
+// - once with dst=NULL, to quickly count the number of Unicode chars,
+// - then with a non-null dst (malloc'ed to the previously obtained size)
+//   to decode and fill it with the Unicode chars.
+
+// adapted from crengine/src/lvstring.cpp
+#define HEAD_CHECK(mask, expect) ((s[0] & mask) == expect)
+#define HEAD_BYTE(mask, shift) (((uint32_t)(s[0]) & mask) << shift)
+#define CONT_BYTE(index, shift) (((uint32_t)(s[index]) & 0x3F) << shift)
+#define HAS_FOLLOWUP(n) (s+n < ends)
+#define IS_FOLLOWING(index) ((s[index] & 0xC0) == 0x80)
+int Utf8ToUnicode(const char * src,  int srclen, uint32_t * dst, int dstlen, bool &is_valid, bool &has_rtl)
```

Comment:
I'd much rather use a fast & accurate [UTF-8 decoder](https://github.com/NiLuJe/FBInk/blob/master/cutef8/dfa.c) (both here and in CRe), but that would obviously throw WTF-8 out the window ;).

Reply:
I thought about asking you to provide me one :) but did not want to get stuck too long with discussion about it :)
But indeed, WTF-8 + a way to plug in RTL char detection is a requirement. And 3 invalid bytes > 3 replacement chars, and non failure (I may have some latin1 in TextEditor :) or invalid stuff from translators) and don't skip the whole invalid but try with next byte a personal preference.
But when comparing with the ones from Harfbuzz and FriBidi, it's always the same logic. Inline the macros and we get the same code :)

Reply:
Yep, glib and utf8proc use a similar approach, IIRC ;).